shims/spark301/src/main/scala/com/nvidia/spark/rapids/shims/spark301/GpuBroadcastHashJoinExec.scala

```diff
@@ -143,10 +143,15 @@ case class GpuBroadcastHashJoinExec(
     val boundCondition = condition.map(GpuBindReferences.bindReference(_, output))
 
     lazy val builtTable = {
-      // TODO clean up intermediate results...
-      val keys = GpuProjectExec.project(broadcastRelation.value.batch, gpuBuildKeys)
-      val combined = combine(keys, broadcastRelation.value.batch)
-      val ret = GpuColumnVector.from(combined)
+      val ret = withResource(
+        GpuProjectExec.project(broadcastRelation.value.batch, gpuBuildKeys)) { keys =>
+        val combined = GpuHashJoin.incRefCount(combine(keys, broadcastRelation.value.batch))
```

Comment:
Shouldn't incrementing the refcount be an implementation detail of `filterBuiltTableIfNeeded` and only if the table wasn't filtered?  Otherwise it seems like the refcounts get off if it was filtered.

Reply:
No, because the contract with `filterBuiltTableIfNeeded` is that it will close the input table and return an output table that the caller must take ownership of and close as needed.  `combine` does not inc any reference counts when it creates `combined` so if we close it then we end up double closing things. 

In the particular case we have to leak `broadcastRelation.value.batch` because of how broadcast works.  So the best I can do is what I am doing now with closing `keys` on the off chance that they are not just a `BoundReference`, but might actually do some computation and produce a result that we were leaking before, which is what the TODO was.

Reply:
I forgot to add that if you want me to add in comments to describe this I can.

Reply:
I am struggling with why this is so hard to follow.  I added comments below for `ShuffledHashJoin` which hopefully explains what is happening.

```scala
        val builtTable = withResource(ConcatAndConsumeAll.getSingleBatchWithVerification(
          buildIter, localBuildOutput)) { buildBatch: ColumnarBatch =>
          // builtBatch owns some data and will be closed when it goes out of scope.
          withResource(GpuProjectExec.project(buildBatch, gpuBuildKeys)) { keys =>
            // keys conceptually owns some data (although it could have just increased the reference counts for
            // things in builtBatch) and will be closed when it goes out of scope
            val combined = GpuHashJoin.incRefCount(combine(keys, buildBatch))
            // We call incRefCount here because conceptually we want combine to look like it copied its
            // input data, but in reality it just placed columns in a new ColumnarBatch.  We inc the ref counts here
            // So that we can close combined.
            // Conceptually combined will be closed by filterBuiltTableIfNeeded and the returned ColumnarBatch (filtered)
            // will be owned by us and need to be closed.  The reality might be a bit more complex than that.
            // Perhaps it should be called filterBuiltTableIfNeededAndCloseInput
            val filtered = filterBuiltTableIfNeeded(combined)
            combinedSize =
                GpuColumnVector.extractColumns(filtered)
                    .map(_.getBase.getDeviceMemorySize).sum.toInt
            // filtered is then closed after it is transitioned to a Table which increments the reference counts for
            // all of the columns
            withResource(filtered) { filtered =>
              GpuColumnVector.from(filtered)
            }
          }
        }
```

The broadcast code looks exactly the same as the above code.  The only difference is that there is no outer `withResources` to close `builtBatch`, because we want to leak the equivalent of `builtbatch`, and there is no `combinedSize` metric because `GpuBroadcastExchange` has a size metric in it. If you want me to make all of that code common I can do it if it will help.

Reply:
> I am struggling with why this is so hard to follow.

It's not you, it's me.  `incRefCount` was added as part of this, and it's weird to me that we've never come across a use-case that needed it before now.  That made me wonder if the refcounting was really correct here and second-guess it.  That, combined with the fact that I didn't fully understand the semantics of `combine` (no pun intended) is what tripped me up.  Apologies for the noise.